klippy/extras/mixingextruder.py

```diff
@@ -0,0 +1,436 @@
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+import math
+import logging
+
+
+def extruder_to_idx(name, active=None):
+    name = name.lower()
+    if name == "active":
+        name = active() if active else ''
+    if name.startswith('mixingextruder'):
+        if name[14:] == '':
+            return 0
+        try:
+            return int(name[14:])
+        except Exception:
+            pass
+    try:
+        return int(name)
+    except Exception:
+        return -1
+
+
+def idx_to_extruder(idx):
+    return "mixingextruder%d" % (idx) if idx else "mixingextruder"
+
+
+def find_mixing_extruder(self, name, active=''):
+    idx = extruder_to_idx(name, lambda: active)
+    return idx_to_extruder(0 if idx < 0 else idx)
+
+
+class MixingMove:
+    def __init__(self, x, y, z, e,
+                 dist_x, dist_y, dist_z, dist_e,
+                 ratio_x, ratio_y, ratio_z, ratio_e,
+                 distance, acceleration,
+                 velocity_start, velocity_cruise,
+                 time_accel, time_cruise, time_decel):
+        self.axes_d = (dist_x, dist_y, dist_z, dist_e)
+        self.axes_r = (ratio_x, ratio_y, ratio_z, ratio_e)
+        self.move_d = distance
+        self.accel = acceleration
+        self.start_v, self.cruise_v = velocity_start, velocity_cruise
+        self.accel_t, self.cruise_t, self.decel_t = \
+            time_accel, time_cruise, time_decel
+        self.start_pos = (x, y, z, e)
+        self.end_pos = tuple(sum(s) for s in zip(self.start_pos, self.axes_d))
+
+
+class MixingExtruder:
+    def __init__(self, config, idx, parent=None):
+        self.printer = config.get_printer()
+        self.activated = False
+        self.name = idx_to_extruder(idx)
+        self.extruder_names = [e.strip()
+                               for e in
+                               config.get('extruders', '').split(",")
+                               if len(e)]
+        if not len(self.extruder_names):
+            raise self.printer.config_error(
+                "No extruders configured for mixing")
+        self.extruders = parent.extruders if parent else []
+        self.mixing_extruders = parent.mixing_extruders if parent else {}
+        self.mixing_extruders[idx] = self
+        self.mixing = self._init_mixings(idx, len(self.extruder_names))
+        self.commanded_pos = 0
+        self.positions = parent.positions if parent else \
+            [0. for p in range(len(self.extruder_names))]
+        self.ratios = [0 for p in range(len(self.extruder_names))]
+        self.current_mixing = tuple(self.ratios)
+        self.gradient_enabled = False
+        # assumed to be sorted list of ((start, middle, end), (ref1, ref2))
+        self.gradients = []
+        self.gradient_method = 'linear'
+        self.printer.register_event_handler("klippy:connect",
+                                            self.handle_connect)
+        logging.info("MixingExtruder %d extruders=%s", idx,
```

Comment:
Missing second `%s` in the format string. Not critical but makes Python throw exceptions.